crates/ty_python_semantic/src/types/generics.rs

```diff
@@ -297,18 +353,25 @@ impl<'db> GenericContext<'db> {
         db: &'db dyn Db,
         types: Box<[Type<'db>]>,
     ) -> Specialization<'db> {
-        assert!(self.variables(db).len() == types.len());
-        Specialization::new(db, self, types, None)
+        assert_eq!(self.variables(db).len(), types.len());
+        debug_assert!(
+            matches!(self.inner(db), GenericContextInner::NonTuple(_)),
+            "Should never call `GenericContext::specialize` on a tuple context"
+        );
```

Comment:
For the record, this is where we had originally put a comment saying "Doug is convinced this will be okay, but Alex is skeptical" :smile: 

So I guess we do need to handle the tuple case here. For that case, we should assert that `types` has one element, and create a `SpecializationInner::Tuple` containing a homogeneous tuple of that type.